go/pkg/diskcache/diskcache.go

```diff
@@ -0,0 +1,331 @@
+				return err
+			}
+			return filepath.WalkDir(prefixDir, func(path string, d fs.DirEntry, err error) error {
+				// We log and continue on all errors, because cache read errors are not critical.
+				if err != nil {
+					return fmt.Errorf("error reading cache directory: %v", err)
+				}
+				if d.IsDir() {
+					return nil
+				}
+				subdir := filepath.Base(filepath.Dir(path))
+				k, err := res.getKeyFromFileName(subdir + d.Name())
+				if err != nil {
+					return fmt.Errorf("error parsing cached file name %s: %v", path, err)
+				}
+				atime, err := getLastAccessTime(path)
+				if err != nil {
+					return fmt.Errorf("error getting last accessed time of %s: %v", path, err)
+				}
+				it := &qitem{
+					key: k,
+					lat: atime,
+				}
+				size, err := res.getItemSize(k)
+				if err != nil {
+					return fmt.Errorf("error getting file size of %s: %v", path, err)
+				}
+				res.store.Store(k, it)
+				atomic.AddInt64(&res.sizeBytes, size)
+				res.mu.Lock()
+				heap.Push(res.queue, it)
+				res.mu.Unlock()
+				return nil
+			})
+		})
+	}
+	if err := eg.Wait(); err != nil {
+		return nil, err
+	}
+	go res.gc()
+	return res, nil
+}
+
+func (d *DiskCache) getItemSize(k key) (int64, error) {
+	return k.digest.Size, nil
+}
+
+// Releases resources and terminates the GC daemon. Should be the last call to the DiskCache.
+func (d *DiskCache) Shutdown() {
+	d.shutdown <- true
+}
+
+func (d *DiskCache) TotalSizeBytes() uint64 {
+	return uint64(atomic.LoadInt64(&d.sizeBytes))
+}
+
+func (d *DiskCache) getKeyFromFileName(fname string) (key, error) {
+	pair := strings.Split(fname, ".")
+	if len(pair) != 2 {
+		return key{}, fmt.Errorf("expected file name in the form [ac_]hash/size, got %s", fname)
+	}
+	size, err := strconv.ParseInt(pair[1], 10, 64)
+	if err != nil {
+		return key{}, fmt.Errorf("invalid size in digest %s: %s", fname, err)
+	}
+	dg, err := digest.New(pair[0], size)
+	if err != nil {
+		return key{}, fmt.Errorf("invalid digest from file name %s: %v", fname, err)
+	}
+	return key{digest: dg}, nil
+}
+
+func (d *DiskCache) getPath(k key) string {
+	return filepath.Join(d.root, k.digest.Hash[:2], fmt.Sprintf("%s.%d", k.digest.Hash[2:], k.digest.Size))
+}
+
+func (d *DiskCache) StoreCas(dg digest.Digest, path string) error {
+	if dg.Size > int64(d.maxCapacityBytes) {
+		return fmt.Errorf("blob size %d exceeds DiskCache capacity %d", dg.Size, d.maxCapacityBytes)
+	}
+	it := &qitem{
+		key: key{digest: dg},
+		lat: time.Now(),
+	}
+	it.mu.Lock()
+	defer it.mu.Unlock()
+	_, exists := d.store.LoadOrStore(it.key, it)
+	if exists {
+		return nil
+	}
+	d.mu.Lock()
+	heap.Push(d.queue, it)
+	d.mu.Unlock()
+	if err := copyFile(path, d.getPath(it.key), dg.Size); err != nil {
+		return err
+	}
+	newSize := uint64(atomic.AddInt64(&d.sizeBytes, dg.Size))
+	if newSize > d.maxCapacityBytes {
+		select {
+		case d.gcReq <- atomic.AddUint64(&d.gcTick, 1):
+		default:
+		}
+	}
+	return nil
+}
+
+func (d *DiskCache) gc() {
```

Comment:
I think this implementation may have the following issues:
- if the capacity is not sufficient for the workload, eviction becomes part of every update.
- having three different locks being acquired and released independently is prune to races; e.g. attempting to load an item after it's been deleted from the ordered list but before it's been deleted from the map, will fail to update the index in the list.
- the client must remember to `Shutdown` the cache to avoid resource leaks.

The capacity issue can be mitigated by evicting a percentage of the cache rather than just enough to meet the threshold. This accommodates several new items before triggering another eviction.

The synchronization issue can be traded off with a simpler implementation that uses a single lock. Updating the list and the map should be an atomic operation in which the entire list is shared and therefore is the bottleneck. Actually, hits are bounded by IO anyways. Misses on the other hand can be faster with less locking. Then again, misses are always followed by updates which are bounded by IO. I think IO is more of a bottleneck than synchronization so it's worth going for a simple locking flow.

How bad is it if eviction events were blocking? I think if eviction saturates IO bandwidth it should be the same whether it's async or sync. Maybe blocking would be faster if we consider disk locality. I also think that most users would prefer using a large cache to avoid eviction anyways. Without a background eviction task, the cache would be simpler to work with.